core/app/page.tsx

```diff
@@ -1,23 +1,36 @@
 import { cookies } from "next/headers";
 import { redirect } from "next/navigation";
 
+import { AuthTokenType } from "db/public";
+
 import { LAST_VISITED_COOKIE } from "~/app/components/LastVisitedCommunity/constants";
 import { getPageLoginData } from "~/lib/authentication/loginData";
+import { createRedirectUrl } from "~/lib/redirect";
+
+export default async function Page({
+	searchParams,
+}: {
+	searchParams: Promise<Record<string, string>>;
+}) {
+	const { user, session } = await getPageLoginData();
 
-export default async function Page() {
-	const { user } = await getPageLoginData();
+	const params = await searchParams;
 
 	if (!user) {
-		redirect("/login");
+		redirect(createRedirectUrl("/login", params).toString());
+	}
+
+	if (session.type === AuthTokenType.verifyEmail) {
+		redirect(createRedirectUrl("/verify", params).toString());
 	}
 
 	const cookieStore = await cookies();
 	const lastVisited = cookieStore.get(LAST_VISITED_COOKIE);
 	const communitySlug = lastVisited?.value ?? user.memberships[0]?.community?.slug;
 
 	if (!communitySlug) {
-		redirect("/settings");
+		redirect(createRedirectUrl("/settings", params).toString());
 	}
 
-	redirect(`/c/${communitySlug}/stages`);
+	redirect(createRedirectUrl(`/c/${communitySlug}/stages`, params).toString());
```

Comment:
the regular pages weren't passing on search params. I needed it to so that `?verified=true` could get passed along to arbitrary pages